### PR TITLE
[compression] Add nullability to (generated and manual) bindings

### PR DESCRIPTION
**Pull request by tj-devel709**
This PR aims to bring nullability changes to Compression.
Following the steps [here](https://github.com/xamarin/xamarin-macios/wiki/Nullability):

1. I am adding `nullable enable` to all manual files that are not "API_SOURCES" in src/frameworks.sources and making the required nullability changes
2. Changing all `throw new ArgumentNullException ("object"));` to `ObjCRuntime.ThrowHelper.ThrowArgumentNullException (nameof (object));` for size saving optimization as well to mark that this framework contains nullability changes
3. Changing any `== null` or `!= null` to `is null` and `is not null`
 
---------------
 
Important notes to this PR:
`_stream` is nullable    =>   `EnsureNotDisposed ()` will throw an error if _stream is null
`_buffer` is nullable     =>    `EnsureBufferInitialized ()` will create a new buffer if null
`_inflator` is nullable   =>   `EnsureDecompressionMode ()` will throw an error if _inflator is null
`_deflator` is nullable  =>  `EnsureCompressionMode ()`will throw an error if _deflator is null
- Since the above calls throw indirect exceptions -> we will need to use '!' for the compiler to know that we are checking

I have left `TODO`s on parts that I am questioning

### CODE REVIEW

File: src/Compression/Compression.cs [hunk 1]
```diff
@@ -169,12 +169,17 @@ private async Task FlushAsyncCore (CancellationToken cancellationToken)
...
-					if (flushSuccessful) {
-						await _stream.WriteAsync(_buffer, 0, compressedBytes, cancellationToken).ConfigureAwait (false);
-					}
+					// TODO Flush throws exception if _buffer is null, do we want to
+					// call EnsureBufferInitialized () to create a _buffer if it is null?
+					flushSuccessful = _deflater!.Flush (_buffer!, out compressedBytes);
+
+					EnsureNotDisposed ();
+					// TODO I think we should call EnsureBufferInitialized () for this next call as well
+					await _stream!.WriteAsync(_buffer!, 0, compressedBytes, cancellationToken).ConfigureAwait (false);
```
Comment (tj-devel709) [inline src/Compression/Compression.cs:182]: @mandel-macaque Then there are some Writes and Reads that take the buffer, but once again, the buffer may be null. Should we call EnsureBufferInitialized () and give it an empty buffer?

File: src/Compression/Compression.cs [hunk 2]
```diff
@@ -666,23 +700,31 @@ public async Task CopyFromSourceToDestination ()
 			{
 				_deflateStream.AsyncOperationStarting ();
 				try {
+					_deflateStream.EnsureDecompressionMode ();
 					// Flush any existing data in the inflater to the destination stream.
 					while (true) {
-						int bytesRead = _deflateStream._inflater.Inflate (_arrayPoolBuffer, 0, _arrayPoolBuffer.Length);
+						// TODO Inflate throws exception if _arrayPoolBuffer is null, do we want to
+						// create a new _arrayPoolBuffer if null or throw an exception here?
```
Comment (tj-devel709) [inline src/Compression/Compression.cs:707]: @mandel-macaque similar things happen down here as well with the `_arrayPoolBuffer`.  I am calling exceptions here if null, but we could also initialize the buffer, or just pass it in the method calls and let them throw the exception?

File: src/Compression/TaskToApm.cs [hunk 3]
```diff
@@ -60,9 +60,15 @@ public static TResult End<TResult>(IAsyncResult asyncResult)
             if (asyncResult is TaskAsyncResult twar && twar._task is Task<TResult> task)
             {
                 return task.GetAwaiter().GetResult();
+            } else if (asyncResult is null) {
+                ObjCRuntime.ThrowHelper.ThrowArgumentNullException (nameof (asyncResult));
+            } else {
+                ObjCRuntime.ThrowHelper.ThrowArgumentNullException (nameof (twar._task));
```
Comment (tj-devel709) [inline src/Compression/TaskToApm.cs:66]: I thought these were better to give us more detailed exceptions, but the compiler does not recognize these `ObjCRuntime.ThrowHelper.ThrowArgumentNullException`s as ending the method so I kept the old one at the end of the method...

File: src/Compression/Compression.cs [hunk 4]
```diff
@@ -101,17 +101,17 @@ private void InitializeBuffer ()
 
 		private void EnsureBufferInitialized ()
```
Comment (mandel-macaque) [inline src/Compression/Compression.cs:102]: Please read https://docs.microsoft.com/en-us/dotnet/csharp/nullable-warnings#possible-dereference-of-null specially the section with the following method signature:
```csharp
private static bool IsNotNull([NotNullWhen(true)] object? obj) => obj != null;
```
We should follow the same pattern.

File: src/Compression/Compression.cs [hunk 5]
```diff
@@ -101,17 +101,17 @@ private void InitializeBuffer ()
...
-			if (_buffer == null) {
+			if (_buffer is null) {
 				InitializeBuffer ();
 			}
 		}
 
-		public Stream BaseStream => _stream;
+		public Stream? BaseStream => _stream;
 
 		public override bool CanRead
```
Comment (mandel-macaque) [inline src/Compression/Compression.cs:111]: same.

Comment (mandel-macaque) [inline src/Compression/Compression.cs:111]: same.

File: src/Compression/Compression.cs [hunk 6]
```diff
@@ -121,7 +121,7 @@ public override bool CanRead
 
 		public override bool CanWrite {
```
Comment (mandel-macaque) [inline src/Compression/Compression.cs:122]: same

Comment (mandel-macaque) [inline src/Compression/Compression.cs:122]: same

File: src/Compression/Compression.cs [hunk 7]
```diff
@@ -169,12 +169,17 @@ private async Task FlushAsyncCore (CancellationToken cancellationToken)
...
+
 				do {
 					int compressedBytes;
-					flushSuccessful = _deflater.Flush (_buffer, out compressedBytes);
-					if (flushSuccessful) {
-						await _stream.WriteAsync(_buffer, 0, compressedBytes, cancellationToken).ConfigureAwait (false);
-					}
+					// TODO Flush throws exception if _buffer is null, do we want to
+					// call EnsureBufferInitialized () to create a _buffer if it is null?
+					flushSuccessful = _deflater!.Flush (_buffer!, out compressedBytes);
```
[the remark this answers is not in the document]
Comment (mandel-macaque) [reply src/Compression/Compression.cs:178]: This is not the correct approach. We do not need the ! if we do what https://docs.microsoft.com/en-us/dotnet/csharp/nullable-warnings#possible-dereference-of-null mentions + the EnsureCompressionMode etc.. that will make the code much cleaner, will also remove the need for ! in most places.

[the remark this answers is not in the document]
Comment (mandel-macaque) [reply src/Compression/Compression.cs:178]: This is not the correct approach. We do not need the ! if we do what https://docs.microsoft.com/en-us/dotnet/csharp/nullable-warnings#possible-dereference-of-null mentions + the EnsureCompressionMode etc.. that will make the code much cleaner, will also remove the need for ! in most places.

File: src/Compression/Compression.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Compression/Compression.cs:182]: same as my previous comment.

Comment (mandel-macaque) [reply src/Compression/Compression.cs:182]: same as my previous comment.

**Review by mandel-macaque — CHANGES_REQUESTED** [review]
Please read the docs and use the examples provided to avoid the need of using ! in all the calls.

**Comment by tj-devel709** [comment]
Closing this PR in favor of this one: https://github.com/xamarin/xamarin-macios/pull/14916

**Pull request closed without merging**